lib/IRGen/GenCall.cpp

```diff
@@ -1371,26 +1377,28 @@ static bool doesClangExpansionMatchSchema(IRGenModule &IGM,
 void SignatureExpansion::expandExternalSignatureTypes() {
   assert(FnType->getLanguage() == SILFunctionLanguage::C);
 
-  // Convert the SIL result type to a Clang type.
-  auto clangResultTy =
-    IGM.getClangType(FnType->getFormalCSemanticResult(IGM.getSILModule()));
+  auto SILResultTy = [&]() {
+    if (FnType->getNumResults() == 0)
+      return SILType::getPrimitiveObjectType(IGM.Context.TheEmptyTupleType);
+
+    return SILType::getPrimitiveObjectType(
+        FnType->getSingleResult().getReturnValueType(
+            IGM.getSILModule(), FnType, TypeExpansionContext::minimal()));
+  }();
+
+  // Convert the SIL result type to a Clang type. If this is for a c++
+  // constructor, use 'void' as the return type.
+  auto clangResultTy = IGM.getClangType(
+      forCXXConstructorCall
+          ? SILType::getPrimitiveObjectType(IGM.Context.TheEmptyTupleType)
```

Comment:
There's `CodeGenTypes::arrangeCXXStructorDeclaration`, but that function isn't available in a header file swift has access to.

Reply:
I tried adding public function `arrangeCXXStructorDeclaration` to `CodeGenABITypes.h` so that it can be called to arrange the constructor signature. But that's causing an assertion to fail in `externalizeArguments`:

https://github.com/apple/swift/blob/main/lib/IRGen/GenCall.cpp#L4019

The cast fails because `AddressOnlyCXXClangRecordTypeInfo` is returned instead of a `LoadableTypeInfo`. The following function returns `AddressOnlyCXXClangRecordTypeInfo`:

https://github.com/apple/swift/blob/6dccf3d6679f442cafa9021de50cf677ac727b53/lib/IRGen/GenStruct.cpp#L1334

The type in question is `StructWithCopyConstructorAndValue` in `test/Interop/Cxx/class/type-classification-non-trivial.swift` and the assertion fails when the argument is passed to the constructor of `StructWithCopyConstructorAndSubobjectCopyConstructorAndValue`:

https://github.com/apple/swift/blob/6dccf3d6679f442cafa9021de50cf677ac727b53/test/Interop/Cxx/class/type-classification-non-trivial.swift#L29

I don't see the assertion fail if I revert the changes and stop using `arrangeCXXStructorDeclaration`. The difference is that `arrangeCXXStructorDeclaration` returns `ABIArgInfo::Indirect` for the argument whereas the previous approach (which passes a pointer type to `arrangeFreeFunctionCall` as the parameter type) returned  `ABIArgInfo::Direct`.

Reply:
It looks like`AddressOnlyCXXClangRecordTypeInfo` has methods like `initializeWithCopy`, which can possibly be used to initialize the temporary.

Reply:
That would introduce an extra copy, which we generally don't want to do in IRGen as part of making a call.

In this case, SIL type lowering already recognized that the type has to be passed indirectly as a parameter or return value and encoded that using an indirect SIL convention.  You should be able to assert that indirect SIL parameters and results correspond to `ABIArgInfo::Indirect` in the Clang `FunctionInfo`.  (The reverse isn't always true.)